src/lib/optimizer/join_graph.cpp

```diff
@@ -0,0 +1,211 @@
+
+const JoinGraph::Vertices& JoinGraph::vertices() const { return _vertices; }
+
+const JoinGraph::Edges& JoinGraph::edges() const { return _edges; }
+
+void JoinGraph::print(std::ostream& out) const {
+  out << "==== JoinGraph ====" << std::endl;
+  out << "==== Vertices ====" << std::endl;
+  for (size_t vertex_idx = 0; vertex_idx < _vertices.size(); ++vertex_idx) {
+    const auto& vertex = _vertices[vertex_idx];
+    std::cout << vertex_idx << ":  " << vertex.node->description() << std::endl;
+  }
+  out << "==== Edges ====" << std::endl;
+  for (const auto& edge : _edges) {
+    if (edge.join_mode == JoinMode::Inner) {
+      std::cout << edge.vertex_indices.first << " <-- " << edge.column_ids->first << " "
+                << scan_type_to_string.left.at(*edge.scan_type) << " " << edge.column_ids->second << " --> "
+                << edge.vertex_indices.second << std::endl;
+    } else {
+      std::cout << edge.vertex_indices.first << " <----> " << edge.vertex_indices.second << std::endl;
+    }
+  }
+
+  out << "===================" << std::endl;
+}
+
+void JoinGraph::_traverse_ast_for_join_graph(const std::shared_ptr<AbstractASTNode>& node,
+                                             JoinGraph::Vertices& o_vertices, JoinGraph::Edges& o_edges) {
+  /**
+   * Early return to make it possible to call search_join_graph() on both children without having to check whether they
+   * are nullptr.
+   */
+  if (!node) {
+    return;
+  }
+
+  Assert(node->num_parents() <= 1, "Nodes with multiple parents not supported when building JoinGraph");
+
+  if (node->type() == ASTNodeType::Join) {
+    const auto join_node = std::static_pointer_cast<JoinNode>(node);
+
+    if (join_node->join_mode() == JoinMode::Inner) {
+      _traverse_inner_join_node(join_node, o_vertices, o_edges);
+    } else if (join_node->join_mode() == JoinMode::Cross) {
+      _traverse_cross_join_node(join_node, o_vertices, o_edges);
+    } else {
+      o_vertices.emplace_back(JoinVertex(node));
+      return;
+    }
+  } else if (node->type() == ASTNodeType::Predicate) {
+    const auto predicate_node = std::static_pointer_cast<PredicateNode>(node);
+
+    // A non-column-to-column Predicate becomes a vertex
+    if (predicate_node->value().type() != typeid(ColumnID)) {
+      _traverse_value_predicate_node(predicate_node, o_vertices, o_edges);
+    } else {
+      _traverse_column_predicate_node(predicate_node, o_vertices, o_edges);
+    }
+  } else {
+    // Everything that is not a Join or a Predicate becomes a vertex
+    o_vertices.emplace_back(JoinVertex(node));
+    return;
+  }
+}
+
+void JoinGraph::_traverse_inner_join_node(const std::shared_ptr<JoinNode>& node, JoinGraph::Vertices& o_vertices,
+                                          JoinGraph::Edges& o_edges) {
+  // The first vertex in o_vertices that belongs to the left subtree.
+  const auto first_left_subtree_vertex_idx = make_join_vertex_id(o_vertices.size());
```

Comment:
Accessing `o_vertices.size()` sounds like accessing the last vertex, not the first.

Reply:
In addition, `first_right_subtree_vertex_idx` in line 116 should contain the exact same value. In combination with the comment, this should confuse the reader